framework/include/userobject/ElementSubdomainModifier.h

```diff
@@ -0,0 +1,95 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ElementUserObject.h"
+#include "NonlinearSystemBase.h"
+#include "AuxiliarySystem.h"
+
+class ElementSubdomainModifier : public ElementUserObject
+{
+public:
+  static InputParameters validParams();
+
+  ElementSubdomainModifier(const InputParameters & parameters);
+
+  virtual void initialSetup() override;
+  virtual void initialize() override;
+  virtual void execute() override;
+  virtual void threadJoin(const UserObject & /*uo*/) override{};
+  virtual void finalize() override;
+
+protected:
+  virtual SubdomainID computeSubdomainID() = 0;
+
+  BoundaryID movingBoundaryID() const
+  {
+    if (!_moving_boundary_specified)
+      mooseError("no moving boundary specified");
+    return _moving_boundary_id;
+  }
+
+  const BoundaryName movingBoundaryName() const
+  {
+    if (!_moving_boundary_specified)
+      mooseError("no moving boundary specified");
+    return _moving_boundary_name;
+  }
+
+  ConstElemRange & movedElemsRange() const { return *_moved_elems_range.get(); }
+
+  ConstBndNodeRange & movedBndNodesRange() const { return *_moved_bnd_nodes_range.get(); }
```

Comment:
I am curious about if we are able to de-reference the object directly from the smart pointer (instead of raw pointer)?

Reply:
You can remove the `.get()`. The dereference operator for a smart pointer returns a reference to the underlying object.

Reply:
You are right, I removed the `.get()`s.